docs/contracts/overview/KeyManager.md

```diff
@@ -99,6 +99,86 @@ Given the example above, the on-chain nonce is 4 and we are executing the relay
+
+```solidity
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.5;
+
+// interfaces
+import {ILSP6KeyManager} from "@lukso/lsp6-contracts/contracts/ILSP6KeyManager.sol";
+
+// modules
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {LSP6SetDataModule} from "@lukso/lsp6-contracts/contracts/LSP6SetDataModule.sol";
+import {LSP6OwnershipModule} from "@lukso/lsp6-contracts/contracts/LSP6OwnershipModule.sol";
+
+/// @title LSP6 Key Manager implementation to enable multiple owners with different permissions and roles
+/// to control an LSP7 or LSP8 Token (instead of having a single `owner()`).
+contract LSP6TokenManager is
+    ILSP6KeyManager,
+    ERC165,
+    LSP6SetDataModule,
+    LSP6OwnershipModule
+{
+    using Address for *;
+    using ECDSA for *;
+    using LSP6Utils for *;
+
+    /// @dev address of the LSP7/8 Token contract this Key Manager controls
+    address private immutable _linkedToken;
+
+    mapping(address => mapping(uint256 => uint256)) internal _nonceStore;
+
+    constructor(address linkedToken_) {
+        if (linkedToken_ == address(0)) revert InvalidLSP6Target();
+        _linkedToken = linkedToken_;
+    }
+
+    /// @dev permission required to update the `LSP4Metadata` data key via `setData(...)` on the token contract
+    bytes32 constant _PERMISSION_UPDATE_TOKEN_METADATA = 0x0000000000000000000000000000000000000000000000000000000000400000;
+
+    /// @dev keccak256('LSP4Metadata') --> from the LSP4 Standard
+    /// This is defined in `LSP4Constants.sol`, but we write it here for better understanding.
+    bytes32 constant _LSP4_METADATA_KEY = 0x9afb95cacc9f95858ec44aa8c3b685511002e30ae54415823f406128b85b238e;
+
+    /// @inheritdoc LSP6SetDataModule
+    /// @dev implement a custom check to verify if the controller
+    /// has the permission to update the token metadata
+    function _getPermissionRequiredToSetDataKey(
+        address controlledContract,
+        bytes32 controllerPermissions,
+        bytes32 inputDataKey,
+        bytes memory inputDataValue
+    ) internal view virtual override returns (bytes32) {
+        if (inputDataKey == _LSP4_METADATA_KEY) {
+            controllerPermissions.hasPermission(
+                _PERMISSION_UPDATE_TOKEN_METADATA
+            );
+        }
+
+        super._getPermissionRequiredToSetDataKey(
+            controlledContract,
+            controllerPermissions,
+            inputDataKey,
+            inputDataValue
+        );
+    }
+
+}
+```
+
+As you can see from the Solidity code snippet above, since the Key Manager is broken down in multiple modules for each types of permissions (`LSP6SetDataModule`, `LSP6OwnershipModule`), it's relatively easy to create a specific implementation by-reusing the same code and implement custom permissions check based on each implementation on top.
```

Comment:
```suggestion
As you can see from the Solidity code snippet above, since the Key Manager is broken down in multiple modules for each types of permissions (`LSP6SetDataModule`, `LSP6OwnershipModule`), it's relatively easy to create a specific implementation by-reusing the same code and implement custom permissions check based on examples above.
```

Reply:
Also, I'm not sure what you meant by `each types of permissions`. It doesn't sound right. Would it be better to use `each set of permissions`?
Or maybe even:
- `each set of permissions that are working in concert`;
- `each set of permissions that are working together to create a specific behaviour`.

Reply:
yeah you are right. It's a set of permissions related to a specific action being performed (setData, execute, etc...)